Doc/reference/datamodel.rst

```diff
@@ -2859,33 +2859,83 @@ through the object's keys; for sequences, it should iterate through the values.
 
 .. note::
 
-   Slicing is done exclusively with the following three methods.  A call like ::
+   Slicing is done exclusively with the following three methods.
+   :meth:`__getitem__`, :meth:`__setitem__`, :meth:`__delitem__`.
+   Expressions like: ::
 
-      a[1:2] = b
+      a[1:2]
+      a[1:2,]
+      a[1:2, 2:3]
 
-   is translated to ::
+   are translated into:. ::
 
-      a[slice(1, 2, None)] = b
+      a[slice(1, 2, None)]
+      a[(slice(1, 2, None),)]
+      a[(slice(1, 2, None), slice(2, 3, None))]
 
-   and so forth.  Missing slice items are always filled in with ``None``.
+   Missing slice items are always filled in with ``None``.
 
 
 .. method:: object.__getitem__(self, key)
 
-   Called to implement evaluation of ``self[key]``. For :term:`sequence` types,
-   the accepted keys should be integers. Optionally, they may support
-   :class:`slice` objects as well.  Negative index support is also optional.
-   If *key* is
-   of an inappropriate type, :exc:`TypeError` may be raised; if *key* is a value
-   outside the set of indexes for the sequence (after any special
-   interpretation of negative values), :exc:`IndexError` should be raised. For
-   :term:`mapping` types, if *key* is missing (not in the container),
-   :exc:`KeyError` should be raised.
+   Called to implement evaluation of ``self[key]``. There are 3 supported
+   duck types: :term:`sequence` types, :term:`mapping` types, and any type
+   with custom behaviour on indexing. Syntax doesn't perform any checks,
+   so it is this method responsibility to raise :exc:`TypeError` on keys of
+   improper type.
+
+   First, there is expected behaviour for :term:`sequence` types:.
+   Minimal requirement for :func:`iter` to work (without existing
+   :meth:`__iter__` method) is that, ``__getitem__`` needs to accept
+   integers from 0 up to the index, on which :exc:`IndexError` is raised,
+   stoping the iteration. :func:`reversed` will also call ``len(self)`` and
+   request only items at non-negative keys. However, if class passes as
+   :class:`collections.abc.Sequence`, ``__getitem__`` should also:.
+
+   1. Accept negative integer keys, for indexing from the end of the sequence.
+   2. For all integer keys outside of bounds ``-len(self) <= key < len(self)``,
```

Comment:
This should not necessarily be the case. For instance, if you implement a list that reduces the index modulo its length.